src/SonarScanner.MSBuild.PreProcessor/WebService/SonarWebService.cs

```diff
@@ -349,43 +244,64 @@ private async Task<T> ExecuteWithLogs<T>(Func<Task<T>> request, Uri logUri)
+            new(serverUri, Escape(query, args));
+
+        protected virtual Uri AddOrganization(Uri uri, string organization) =>
+            string.IsNullOrEmpty(organization)
+                ? uri
+                : new Uri(uri + $"&organization={WebUtility.UrlEncode(organization)}");
+
+        private Dictionary<string, string> ParseSettingsResponse(string contents)
         {
-            var key = p["key"].ToString();
-            if (p["value"] != null)
+            var settings = new Dictionary<string, string>();
+            var settingsArray = JObject.Parse(contents).Value<JArray>("settings");
+            foreach (var t in settingsArray)
             {
-                var value = p["value"].ToString();
-                settings.Add(key, value);
+                GetPropertyValue(settings, t);
             }
-            else if (p["fieldValues"] != null)
+
+            return CheckTestProjectPattern(settings);
+        }
+
+        private IList<SensorCacheEntry> ParseCacheEntries(Task<Stream> task)
+        {
+            if (task.IsFaulted || task.Result is not { } dataStream)
             {
-                MultivalueToProps(settings, key, (JArray)p["fieldValues"]);
+                return new List<SensorCacheEntry>();
             }
-            else if (p["values"] != null)
+
+            var cacheEntries = new List<SensorCacheEntry>();
+            try
             {
-                var array = (JArray)p["values"];
-                var value = string.Join(",", array.Values<string>());
-                settings.Add(key, value);
+                while (dataStream.Position < dataStream.Length)
+                {
+                    cacheEntries.Add(SensorCacheEntry.Parser.ParseDelimitedFrom(dataStream));
+                }
             }
-            else
+            catch (Exception e)
             {
-                throw new ArgumentException("Invalid property");
+                logger.LogDebug(Resources.MSG_IncrementalPRCacheEntryDeserialization, e.Message);
             }
+
+            return cacheEntries;
         }
 
-        private void MultivalueToProps(Dictionary<string, string> props, string settingKey, JArray array)
+        protected static string GetComponentIdentifier(string projectKey, string projectBranch = null)
         {
-            var id = 1;
-            foreach (var obj in array.Children<JObject>())
+            var projectId = projectKey;
+            if (!string.IsNullOrWhiteSpace(projectBranch))
             {
-                foreach (var prop in obj.Properties())
-                {
-                    var key = string.Concat(settingKey, ".", id, ".", prop.Name);
-                    var value = prop.Value.ToString();
-                    props.Add(key, value);
-                }
-                id++;
+                projectId = projectKey + ":" + projectBranch;
             }
+
+            return projectId;
         }
 
         private static string ParseRuleKey(string key)
```

Comment:
This can also be a single line arrow method
```
        private static string ParseRuleKey(string key) =>
            key.Substring(key.IndexOf(':') + 1);
```